[CodeGen][KCFI] Allow setting type hash from xxHash64 to FNV-1a

kees opened the pull request:
Allow for switching the KCFI type id hash from xxHash64 to FNV-1a (to match the coming GCC KCFI implementation). Introduces the Clang `-fsanitize-kcfi-hash=` option and "kcfi-hash" IR module option.

kees committed:
[CodeGen][KCFI] Emit signed comment for .set directive value

When emitting the assembly .set directive, KCFI needs to use
getZExtValue(). However, this means that FileCheck pattern matching can't
match between the .set directive and the IR when the high bit of a 32-bit
value is set. We had gotten lucky with the existing tests happening to
just not have had the high bit set. The coming hash change will expose
this, though.

LLVM IR's default printing behavior uses APInt::operator<<, which calls
APInt::print(OS, /*isSigned=*/true). This means KCFI operand bundles in
call instructions print as signed (e.g. [ "kcfi"(i32 -1208803271) ]),
and KCFI type metadata prints as signed (e.g. !3 = !{i32 -1208803271}).
Changing the IR to print unsigned i32 values would impact hundreds of
existing tests, so it is best to just leave it be.

Update the KCFI .set direct to use getSExtValue() in a comment so that
we can both build correctly and use FileCheck with pattern matching in
tests.

Signed-off-by: Kees Cook <kees@kernel.org>

kees committed:
[CodeGen][KCFI] Introduce llvm::getKCFITypeID to wrap hash implementation

KCFI generates hashes in two places. Instead of exposing the hash
implementation in both places, introduce a helper that wraps the
specific hash implementation in a single place.

Signed-off-by: Kees Cook <kees@kernel.org>

samitolvanen commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -3186,7 +3186,7 @@ void CodeGenModule::finalizeKCFITypes() {
 
     std::string Asm = (".weak __kcfi_typeid_" + Name + "\n.set __kcfi_typeid_" +
-                       Name + ", " + Twine(Type->getZExtValue()) + "\n")
+                       Name + ", " + Twine(Type->getSExtValue()) + "\n")
```
OK, testing with this change alone, x86_64 defconfig build fails when linking .tmp_vmlinux1:
```
ld.lld: error: vmlinux.o:(function __cfi___memcpy: .noinstr.text+0x4a21): relocation R_X86_64_32 out of range: 18446744071952041092 is not in [0, 4294967295]; references '__kcfi_typeid___memcpy'
>>> referenced by usercopy_64.c
>>> defined in vmlinux.o
...
```

kees commented:
Is the version of LLVM used by rustc exposed in some way? I'd like to avoid a flag for this if we can just say "kcfi only works on [combination of clang version and llvm version]". Does that seem feasible?

kees committed:
[CodeGen][KCFI] Provide alternative KCFI hash

In order to transition between KCFI hash, we need to be able to specify
them. Add the Clang option -fsanitize-kcfi-hash= and a IR module option
"kcfi-hash" that can choose between xxHash64 and FNV-1a. Default to
xxHash64 to stay backward compatible, as we'll need to also update rustc
to take a new option to change the hash to FNV-1a for interop with the
coming GCC KCFI.

fmayer commented on llvm/include/llvm/Support/Hash.h:
```diff
@@ -0,0 +1,38 @@
+/// Compute KCFI type ID from mangled type name.
+/// The algorithm can be xxHash64 or FNV-1a.
+uint32_t
+getKCFITypeID(StringRef MangledTypeName,
```
should we leave out the default argument? then we make sure we don't forget to think about the type id anywhere

fmayer reviewed (commented):
Thanks, I had concerns about just changing the type, a flag SGTM

nathanchance commented:
This breaks `ARCH=arm64 allmodconfig`:

```
$ make -skj"$(nproc)" ARCH=arm64 LLVM=1 clean allmodconfig ipc/ipc_sysctl.o
<inline asm>:6:55: error: expected newline
    6 | .set __kcfi_typeid_proc_doulongvec_minmax, 3439065979 # -855901317
      |                                                       ^
<inline asm>:8:53: error: expected newline
    8 | .set __kcfi_typeid_proc_dointvec_minmax, 3439065979 # -855901317
      |                                                     ^
2 errors generated.
make[5]: *** [scripts/Makefile.build:287: ipc/ipc_sysctl.o] Error 1
...

kees commented:
> This breaks `ARCH=arm64 allmodconfig`:
> 
> ```
> $ make -skj"$(nproc)" ARCH=arm64 LLVM=1 clean allmodconfig ipc/ipc_sysctl.o
> <inline asm>:6:55: error: expected newline
>     6 | .set __kcfi_typeid_proc_doulongvec_minmax, 3439065979 # -855901317
>       |                                                       ^
> <inline asm>:8:53: error: expected newline
>     8 | .set __kcfi_typeid_proc_dointvec_minmax, 3439065979 # -855901317
>       |                                                     ^
> 2 errors generated.
> make[5]: *** [scripts/Makefile.build:287: ipc/ipc_sysctl.o] Error 1
> ...
> ```

Argh. Does aarch64 use a different asm comment character?
